src/ray/core_worker/core_worker.h

```diff
@@ -222,8 +223,14 @@ class CoreWorker {
   /// a corresponding `Create()` call and then writing into the returned buffer.
   ///
   /// \param[in] object_id Object ID corresponding to the object.
+  /// \param[in] owns_object Whether or not this worker owns the object. If true,
+  ///            the object will be added as owned to the reference counter as an
+  ///            owned object and this worker will be responsible for managing its
+  ///            lifetime.
+  /// \param[in] pin_object Whether or not to pin the object at the local raylet. This
+  ///            only applies when owns_object is true.
   /// \return Status.
-  Status Seal(const ObjectID &object_id);
+  Status Seal(const ObjectID &object_id, bool owns_object, bool pin_object);
```

Comment:
I'm trying to understand better on the flags. So for `Put` we always pin the object as it's owned by us. For `Seal` we still pin the objects for non-tests scenarios, where we own the object. It's only for tests that user can pass in an object id and in that case we won't pin the object. Is this the case? @edoakes

Reply:
I think one point of confusion is that `Put` is actually not used by Python at all and `Seal` is not used by Java at all - they each use the respective one for `ray.put()` calls. I only added this complex logic for `Seal` here because it's only used in Python tests but could add it for `Put` as well if you think that's better. Otherwise your summary is correct.

Reply:
I see, thanks. Yes that caused my confusion. I think it would be nice to also have it in java, so the features are more consistent for different user languages. If it's too late for this PR, it's also fine to do it in a later one.